libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/ServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,129 @@
+        public static void AddBotRuntime<TBot>(this IServiceCollection services, IConfiguration configuration) 
+            where TBot : class, IBot
+        {
+            RegisterCommonServices(services, configuration);
+
+            services.AddSingleton<IBot, TBot>();
+        }
+
+        /// <summary>
+        /// Adds bot related services to the application's service collection.
+        /// </summary>
+        /// <remark>
+        /// The following dependencies are added with TrySingleton so advanced scenarios can override them to customize the runtime behavior:
+        /// <see cref="BotFrameworkAuthentication"/>,
+        /// <see cref="IBotFrameworkHttpAdapter"/>,
+        /// <see cref="IStorage"/>,
+        /// <see cref="UserState"/>,
+        /// <see cref="ConversationState"/>
+        /// 
+        /// This set of dependencies is designed to be sufficient to run a typical bot. Since each of these
+        /// are registered using TrySingleton if you want to provide a different implementation of any of the dependencies
+        /// just register them with the service collection before calling AddBotRuntime.
+        /// 
+        /// </remark>
+        /// <param name="services">The application's collection of registered services.</param>
+        /// <param name="configuration">The application configuration.</param>
+        /// <typeparam name="TBot">Type of Bot.</typeparam>
+        /// <typeparam name="TDialog">Type of Dialog.</typeparam>
+        public static void AddBotRuntime<TBot, TDialog>(this IServiceCollection services, IConfiguration configuration)
+            where TBot : class, IBot
+            where TDialog : Dialog
+        {
+            RegisterCommonServices(services, configuration);
+            services.AddSingleton<TDialog>();
+            services.AddSingleton<IBot, TBot>();
+        }
+
+        private static void RegisterCommonServices(IServiceCollection services, IConfiguration configuration)
+        {
+            if (services == null)
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
+            // Ensure the IConfiguration is available. (Azure Functions don't do this.)
+            services.TryAddSingleton(configuration);
+
+            // Add basic authentication.
+            services.TryAddSingleton<BotFrameworkAuthentication, ConfigurationBotFrameworkAuthentication>();
+
+            // Add a CoreBotAdapter as the IBotFrameworkHttpAdapter unless one was already registered.
+            services.TryAddSingleton<IBotFrameworkHttpAdapter, CoreBotAdapter>();
+
+            // Create the storage we'll be using for User and Conversation state. (Memory is great for testing purposes.)
+            services.TryAddSingleton<IStorage>(new MemoryStorage());
```

Comment:
Memory should only be used for dev and tests, we should find a way of making obvious to the dev that they are using MemoryStorage and that they MUST opt for cosmos or azure storage for prod bots.